<a href="https://colab.research.google.com/github/priyanshu7466/Python/blob/main/ML/Grid_Search.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
iris = load_iris()

In [2]:
from sklearn.svm import SVC
svm = SVC()

In [3]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(iris.data,iris.target,test_size=0.3)

In [4]:
len(X_train)

105

In [5]:
from sklearn.model_selection import cross_val_score

In [6]:
cross_val_score(SVC(kernel='rbf',C=10),X_train,y_train,cv=5)

array([1.        , 0.95238095, 1.        , 0.95238095, 1.        ])

In [7]:
cross_val_score(SVC(kernel='linear',C=10),X_train,y_train,cv=5)

array([1.        , 0.95238095, 0.95238095, 1.        , 1.        ])

In [8]:
cross_val_score(SVC(kernel='rbf',C=20),X_train,y_train,cv=5)

array([1.        , 0.95238095, 1.        , 0.95238095, 1.        ])

In [13]:
kernels = ['rbf','linear']
c = [1,10,20]
scores = {}
for kval in kernels:
  for cval in c:
    cv_scores = cross_val_score(SVC(kernel=kval,C=cval,gamma='auto'),X_train,y_train,cv=5)
    scores[kval + '_' + str(cval)] = np.average(cv_scores)
scores

{'rbf_1': 0.980952380952381,
 'rbf_10': 0.980952380952381,
 'rbf_20': 0.9714285714285713,
 'linear_1': 0.9904761904761905,
 'linear_10': 0.980952380952381,
 'linear_20': 0.9714285714285713}

In [17]:
from sklearn.model_selection import GridSearchCV
clf = GridSearchCV(SVC(gamma='auto'),{
    'C' : [1,10,20],
    'kernel' : ['rbf','linear']
},cv=5,return_train_score=False)
clf.fit(X_train,y_train)
clf.cv_results_

{'mean_fit_time': array([0.00126343, 0.00081935, 0.00088639, 0.00076008, 0.00079851,
        0.00082827]),
 'std_fit_time': array([4.18454381e-04, 6.32812675e-05, 4.53648347e-05, 2.96560513e-05,
        1.43902155e-05, 1.22460379e-04]),
 'mean_score_time': array([0.00085106, 0.00056219, 0.00063496, 0.00054312, 0.00054841,
        0.00066342]),
 'std_score_time': array([2.83000166e-04, 1.18453491e-05, 8.33931344e-05, 8.79373567e-06,
        1.14593801e-05, 1.13736868e-04]),
 'param_C': masked_array(data=[1, 1, 10, 10, 20, 20],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_kernel': masked_array(data=['rbf', 'linear', 'rbf', 'linear', 'rbf', 'linear'],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'C': 1, 'kernel': 'rbf'},
  {'C': 1, 'kernel': 'linear'},
  {'C': 10, 'kernel': 'rbf'},
  {'C': 10, 'kernel': 'linear'},
  {'C': 20, 'ker

In [18]:
df = pd.DataFrame(clf.cv_results_)
df

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0       0.001263      0.000418         0.000851        0.000283       1   
1       0.000819      0.000063         0.000562        0.000012       1   
2       0.000886      0.000045         0.000635        0.000083      10   
3       0.000760      0.000030         0.000543        0.000009      10   
4       0.000799      0.000014         0.000548        0.000011      20   
5       0.000828      0.000122         0.000663        0.000114      20   

  param_kernel                         params  split0_test_score  \
0          rbf      {'C': 1, 'kernel': 'rbf'}           1.000000   
1       linear   {'C': 1, 'kernel': 'linear'}           1.000000   
2          rbf     {'C': 10, 'kernel': 'rbf'}           1.000000   
3       linear  {'C': 10, 'kernel': 'linear'}           1.000000   
4          rbf     {'C': 20, 'kernel': 'rbf'}           0.952381   
5       linear  {'C': 20, 'kernel': 'linear'}           0.952381   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.952381           0.952381           1.000000                1.0   
1           0.952381           1.000000           1.000000                1.0   
2           0.952381           1.000000           0.952381                1.0   
3           0.952381           0.952381           1.000000                1.0   
4           0.952381           0.952381           1.000000                1.0   
5           0.952381           0.952381           1.000000                1.0   

   mean_test_score  std_test_score  rank_test_score  
0         0.980952        0.023328                2  
1         0.990476        0.019048                1  
2         0.980952        0.023328                2  
3         0.980952        0.023328                2  
4         0.971429        0.023328                5  
5         0.971429        0.023328                5

In [19]:
df[['param_C','param_kernel','mean_test_score']]

param_C param_kernel  mean_test_score
0       1          rbf         0.980952
1       1       linear         0.990476
2      10          rbf         0.980952
3      10       linear         0.980952
4      20          rbf         0.971429
5      20       linear         0.971429

In [20]:
clf.best_params_

{'C': 1, 'kernel': 'linear'}

In [21]:
clf.best_score_

0.9904761904761905

In [22]:
dir(clf)

['__abstractmethods__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_check_feature_names',
 '_check_n_features',
 '_check_refit_for_multimetric',
 '_estimator_type',
 '_format_results',
 '_get_param_names',
 '_get_tags',
 '_more_tags',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_required_parameters',
 '_run_search',
 '_select_best_index',
 '_validate_data',
 '_validate_params',
 'best_estimator_',
 'best_index_',
 'best_params_',
 'best_score_',
 'classes_',
 'cv',
 'cv_results_',
 'decision_function',
 'error_score',
 'estimator',
 'fit',
 'get_params',
 'inverse_transform',
 'multimetric_',
 'n_fe

In [25]:
from sklearn.model_selection import RandomizedSearchCV
rs = RandomizedSearchCV(SVC(gamma='auto'),{
    'C' : [1,10,20],
    'kernel' : ['rbf','linear']
},cv=5,return_train_score=False,n_iter=2)
rs.fit(X_train,y_train)
pd.DataFrame(rs.cv_results_)[['param_C','param_kernel','mean_test_score']]

param_C param_kernel  mean_test_score
0      20          rbf         0.971429
1       1       linear         0.990476

In [41]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

model_params = {
    'svm' : {
        'model' : SVC(gamma='auto'),
        'params' : {
          'C' : [1,10,20],
          'kernel' : ['rbf','linear']
      }
    },
    'random_forest' : {
        'model' : RandomForestClassifier(),
        'params' : {
            'n_estimators' : [1,5,10]
        }
    },
    'logistic_regression' : {
        'model' : LogisticRegression(),
        'params' : {
            'C' : [1,5,10],
            'solver':['liblinear']
        }
    }
}

In [42]:
scores = []

for model_name, mp in model_params.items():
  clf = GridSearchCV(mp['model'],mp['params'],cv=5,return_train_score=False)
  clf.fit(X_train,y_train)
  scores.append({
      'model' : model_name,
      'best_score' : clf.best_score_,
      'best_params' : clf.best_params_
  })

df = pd.DataFrame(scores,columns=['model','best_score','best_params'])
df

model  best_score                      best_params
0                  svm    0.990476     {'C': 1, 'kernel': 'linear'}
1        random_forest    0.961905             {'n_estimators': 10}
2  logistic_regression    0.961905  {'C': 5, 'solver': 'liblinear'}

Based on above, I can conclude that SVM with C=1 and kernel='linear' is the best model for solving my problem of iris flower classification

# Exercise

In [29]:
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.tree import DecisionTreeClassifier

In [30]:
from sklearn.datasets import load_digits
digits = load_digits()

In [38]:
model_params = {
    'svm' : {
        'model' : SVC(gamma='auto'),
        'params' : {
            'C' : [1,10,20,50,70],
            'kernel': ['rbf','linear']
        }
    },
    'random_forest' : {
        'model': RandomForestClassifier(),
        'params':{
            'n_estimators': [1,3,5,7,10,20]
        }
    },
    'logistic_regression':{
        'model': LogisticRegression(),
        'params':{
            'C': [1,3,5,7,10,20],
            'solver': ['liblinear']
        }
    },
    'gaussian':{
        'model': GaussianNB(),
        'params': {

        }
    },
    'multinomial':{
        'model': MultinomialNB(),
        'params': {

        }
    },
    'decision_tree':{
        'model': DecisionTreeClassifier(),
        'params': {
            'criterion': ['gini','entropy']
        }
    }
}

In [39]:
scores = []

for model_name, mp in model_params.items():
  clf = GridSearchCV(mp['model'],mp['params'],cv=5,return_train_score=False)
  clf.fit(digits.data,digits.target)
  scores.append({
      'model': model_name,
      'params' : clf.best_params_,
      'best_score': clf.best_score_
  })

df = pd.DataFrame(scores,columns=['model','params','best_score'])
df

model                           params  best_score
0                  svm     {'C': 1, 'kernel': 'linear'}    0.947697
1        random_forest             {'n_estimators': 20}    0.919896
2  logistic_regression  {'C': 1, 'solver': 'liblinear'}    0.922114
3             gaussian                               {}    0.806928
4          multinomial                               {}    0.870350
5        decision_tree         {'criterion': 'entropy'}    0.805272

For me the winner is svm (C=1, kernel=linear) with 94.77% score. It could be different for you as I have limited my parameters to be certain values only